lib/phoenix_live_view/engine.ex

```diff
@@ -984,20 +1040,39 @@ defmodule Phoenix.LiveView.Engine do
     {expr, vars, assigns}
   end
 
-  # Vars always taint unless we are in restricted mode.
-  defp analyze({name, meta, nil} = expr, {:restricted, map}, assigns, caller)
+  # Vars always taint unless we are in restricted mode
+  # or the variable is marked as `:change_track` for vars_changed.
+  defp analyze({name, meta, nil} = expr, {:restricted, map} = vars, assigns, caller)
        when is_atom(name) do
-    if Map.has_key?(map, name) do
-      maybe_warn_taint(name, meta, caller)
-      {expr, {:tainted, map}, assigns}
-    else
-      {expr, {:restricted, map}, assigns}
+    case map do
+      %{^name => :tainted} ->
+        maybe_warn_taint(name, meta, caller)
+        {expr, {:tainted, map}, assigns}
+
+      %{^name => :change_track} ->
+        {expr, vars, Map.put(assigns, {:vars_changed, [name]}, true)}
+
+      _ ->
+        {expr, {:restricted, map}, assigns}
     end
   end
 
-  defp analyze({name, meta, nil} = expr, {_, map}, assigns, caller) when is_atom(name) do
-    maybe_warn_taint(name, meta, caller)
-    {expr, {:tainted, Map.put(map, name, true)}, assigns}
+  defp analyze({name, meta, nil} = expr, {type, map}, assigns, caller)
+       when is_atom(name) do
+    cond do
+      Map.get(map, name) == :change_track ->
+        {expr, {type, map}, Map.put(assigns, {:vars_changed, [name]}, true)}
+
+      :change_track in meta ->
```

Comment:
meta should be a keyword list, so let's make sure to store `{:change_track, true}` in there instead.